### PR TITLE
auto-update flow

### PR DESCRIPTION
**Pull request by LukeSheard**
Add a prompt during startup to validate whether users wish to update their version of modular when there is a new one available in the registry their repository is aligned with. Added wrapper around fetch to deal with HTTP / HTTPS proxies.

Screenshot 2020-12-23 at 6 54 41 pm

### CODE REVIEW

File: packages/modular-scripts/src/cli.ts [hunk 1]
```diff
@@ -59,9 +62,7 @@ function isYarnInstalled(): boolean {
 
 type PackageType = 'app' | 'view' | 'root'; // | 'package', the default
 
-export interface PackageJson {
-  name: string;
-  private?: boolean;
+export type ModularPackageJson = PackageJson & {
```
Comment (threepointone) [inline packages/modular-scripts/src/cli.ts:65]: I can't recollect why this was previously exported, and I can't find an import for this type elsewhere anywhere now. Remove this export?

File: packages/modular-scripts/src/fetch.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  if (process.env.HTTP_PROXY || process.env.HTTPS_PROXY) {
+    return proxyAgent;
+  }
+};
+const proxyAgent = (parsedUrl: URL) => {
+  const protocol = parsedUrl.protocol;
+  if (protocol === 'https') {
+    return new HttpsProxyAgent(process.env.HTTPS_PROXY as string);
+  } else {
+    return new HttpsProxyAgent(process.env.HTTP_PROXY as string);
```
Comment (threepointone) [inline packages/modular-scripts/src/fetch.ts:14]: sorry for my noob question, but will this work with work on http? There appears to be an `http-proxy-agent` module, should you be using that?

File: packages/modular-scripts/src/cli.ts [hunk 3]
```diff
@@ -343,6 +339,44 @@ async function build(directoryName: string, preserveModules?: boolean) {
...
+    const newVersion = modualarScriptsRegistry["dist-tags"]["latest"];
+
+    const { version }: ModularPackageJson = fs.readJSONSync(path.join(__dirname, "..", "package.json"));
+    if (newVersion != version) {
+      const { dependencies }: ModularPackageJson = fs.readJSONSync(path.join(getModularRoot(), "package.json"));
+
+      const update  = await prompts({
+        type: 'confirm',
+        name: 'value',
+        message: `Modular is out of date - would you like to update from ${version} to ${newVersion}`,
```
Comment (threepointone) [inline packages/modular-scripts/src/cli.ts:363]: `modular` is lowercase. Also a suggestion for the wording. 
```suggestion
        message: `Your version of modular is ${version} . Would you like to update to ${newVersion}?`,
```

File: packages/modular-scripts/src/cli.ts [hunk 4]
```diff
@@ -343,6 +339,44 @@ async function build(directoryName: string, preserveModules?: boolean) {
...
+  if (process.env.SKIP_PREFLIGHT_CHECK !== 'true') {
+    const { stdout: registry } = await execa('yarnpkg', ['config', 'get', 'registry']);
+
+    const url = String(new URL("/modular-scripts", registry));
+
+    const res = await fetch(url);
+    const modualarScriptsRegistry: NPMRegistryListing = await res.json();
+    const newVersion = modualarScriptsRegistry["dist-tags"]["latest"];
+
+    const { version }: ModularPackageJson = fs.readJSONSync(path.join(__dirname, "..", "package.json"));
```
Comment (threepointone) [inline packages/modular-scripts/src/cli.ts:356]: This doesn't seem right, this will only work in our own repository (which is... odd, what does updating mean then). Did you mean to use dependencies['modular-scripts'] and compare with `semver` instead?

Also, instead of using package.json, maybe you should be using yarn.lock instead?

File: packages/modular-scripts/src/cli.ts [hunk 5]
```diff
@@ -343,6 +339,44 @@ async function build(directoryName: string, preserveModules?: boolean) {
...
+
+async function preflight() {
+  if (process.env.SKIP_PREFLIGHT_CHECK !== 'true') {
+    const { stdout: registry } = await execa('yarnpkg', ['config', 'get', 'registry']);
+
+    const url = String(new URL("/modular-scripts", registry));
+
+    const res = await fetch(url);
+    const modualarScriptsRegistry: NPMRegistryListing = await res.json();
+    const newVersion = modualarScriptsRegistry["dist-tags"]["latest"];
```
Comment (threepointone) [inline packages/modular-scripts/src/cli.ts:354]:
```suggestion
    const modularScriptsRegistry: NPMRegistryListing = await res.json();
    const newVersion = modularScriptsRegistry["dist-tags"]["latest"];
```

**Review by threepointone — COMMENTED** [review]
A problem here is that it'll bother people to update every time they run, and that could be annoying. Also you'd have to disable this command on CI. Instead, how about just logging the message to update (with a copy paste command `yarn upgrade modular-scripts eslint-config-modular-app`, and let them do it whenever they want?

File: packages/modular-scripts/src/cli.ts (hunk 4 above)
Comment (threepointone) [reply packages/modular-scripts/src/cli.ts:356]: Oh it'll look at the package.json of itself. I think you should use find-up to find the nearest package.json instead, because you can't guarantee that this file will be at this path (especially after we migrate to using `modular build` to build modular-scripts)

**Comment by threepointone** [comment]
This is super impactful, wish we'd done it from the beginning. Thanks for the PR on your time off!
Trying to think of how we can write a tests for this 🤔

File: packages/modular-scripts/src/fetch.ts (hunk 2 above)
Comment (LukeSheard) [reply packages/modular-scripts/src/fetch.ts:14]: lol this is my noob mistake - I wish there was an abstracted version of this as a package. I will just take this for inspiration... https://github.com/touv/node-fetch-with-proxy/blob/master/src/index.js#L53

File: packages/modular-scripts/src/cli.ts (hunk 1 above)
Comment (LukeSheard) [reply packages/modular-scripts/src/cli.ts:65]: 👍

File: packages/modular-scripts/src/cli.ts (hunk 4 above)
Comment (LukeSheard) [reply packages/modular-scripts/src/cli.ts:356]: If I can work out jest mocks then it should be easy enough to mock out the `fetch` and the `fse` call to return differing things, then just assert that prompts was called?

**Comment by LukeSheard** [comment]
> Trying to think of how we can write a tests for this 🤔

If I can work out jest mocks then it should be easy enough to mock out the `fetch` and the `fse` call to return differing things, then just assert that prompts was called?

File: packages/modular-scripts/src/cli.ts (hunk 4 above)
Comment (threepointone) [reply packages/modular-scripts/src/cli.ts:356]: Mocking fetch is a dope idea. Yeah I think you're right, that should do it.

**Comment by threepointone** [comment]
Mocking fetch is a dope idea. Yeah I think you're right, that should do it.

File: packages/modular-scripts/src/cli.ts (hunk 4 above)
Comment (LukeSheard) [reply packages/modular-scripts/src/cli.ts:356]: I went around in circles a couple of times around this. There's also the question of how we deal with a major releases (once we get to `1.0.0`) and make that transition easy for people - although this preflight would give us the option to run codemods on install as well automatically.

**Comment by LukeSheard** [comment]
> A problem here is that it'll bother people to update every time they run, and that could be annoying.

I think the easiest thing to do on this front is to add a `last check time` and only run once a (day|week) so that we don't spam too hard. Although I think we should push people to be on the latest version as much as possible as upgrading should be minimal effort and they get dope features. 💯

**Comment by NMinhNguyen** [comment]
Is there any reason not to use a package like https://github.com/yeoman/update-notifier instead? It's what `npm` CLI uses.

**Comment by threepointone** [comment]
>  Although I think we should push people to be on the latest version as much as possible as upgrading should be minimal effort and they get dope features. 💯

I agree in principle, but consider the workflow. They'll see this prompt while they're working on other stuff, and won't be able to update immediately. It'll likely happen in a separate PR where they can isolate this upgrade. Also, if we have breaking changes, it won't be just the package that has to be updated, they'll have to make some other changes too. For this to happen safely we'll require at the minimum a `modular check` command, if not some form of automated codemod/filemod process (related issue: https://github.com/jpmorganchase/modular/issues/110). So until that happens, I don't think it should be a prompt. 

In general we shouldn't deprecate versions if it's working fine (Imagine if yarn 1 kept asking us to update to yarn 2. I'd rather not use yarn at all then.). It's a different issue *inside* JP Morgan, and we'll pull different levers there to get teams to update (One thing to do is to literally update it ourselves haha)

To summarise, I don't think we should have the prompt _now_, but should *definitely* have it at some point when we have the full upgrade story in place. Fair?

**Comment by tstordyallison** [comment]
Maybe it’s more of a single printed line saying it’s out of date + a hint to run some command to easy update? And it could show/check only once a day etc. 

Then it’s less in your face/blocking the run of the command you trying do right now. But a gentle prod to run the command when it’s a good time to.

**Comment by threepointone** [comment]
Yup. To clarify, I meant we shouldn't have the "prompts()" call, but definitely have the message and copy-pastable command.

**Comment by threepointone** [comment]
It can even show on every run then, that's not as big a problem.

File: packages/modular-scripts/src/preflight.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+async function isYarnInstalled(): Promise<boolean> {
+  try {
+    await execa('yarnpkg', ['-v']);
+    return true;
+  } catch (err) {
+    return false;
+  }
+}
+
+async function preflight(): Promise<void> {
```
Comment (threepointone) [inline packages/modular-scripts/src/preflight.ts:17]: could you call this (and the filename) `preflightCheck` instead? preflight is an adjective, not a good function name.

File: packages/modular-scripts/package.json [hunk 7]
```diff
@@ -9,6 +9,7 @@
     "node": ">=12"
   },
   "scripts": {
+    "modular": "ts-node src/cli.ts",
```
Comment (threepointone) [inline packages/modular-scripts/package.json:12]: why?

File: yarn.lock [hunk 8]
```diff
@@ -1765,6 +1765,11 @@
   resolved "https://registry.yarnpkg.com/@schemastore/tsconfig/-/tsconfig-0.0.9.tgz#77a1a1fb55eb879391aa78db3154dbcc4d93948b"
```
Comment (threepointone) [inline yarn.lock:1765]: wow, this is a lot of dependencies. wild.

File: packages/modular-scripts/src/preflight.ts [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+  if (process.env.SKIP_PREFLIGHT_CHECK !== 'true') {
+    const { name, version } = fs.readJSONSync(
+      path.join(__dirname, '..', 'package.json'),
+    ) as PackageJson;
+
+    const message =
+      'Modular is out of date - you can update from ' +
```
Comment (threepointone) [inline packages/modular-scripts/src/preflight.ts:24]: modular is lowercase

File: packages/modular-scripts/src/preflight.ts [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+  try {
+    await execa('yarnpkg', ['-v']);
+    return true;
+  } catch (err) {
+    return false;
+  }
+}
+
+async function preflight(): Promise<void> {
+  if (process.env.SKIP_PREFLIGHT_CHECK !== 'true') {
```
Comment (threepointone) [inline packages/modular-scripts/src/preflight.ts:18]: disable this when CI=true as well

**Comment by threepointone** [comment]
got any ideas on how to pass the CI failure?

File: yarn.lock (hunk 8 above)
Comment (LukeSheard) [reply yarn.lock:1765]: Yeah I know - the world of 1 liner modules...

**Comment by LukeSheard** [comment]
I'm not sure how we fix the CI since the error is coming from the fact we use `isolatedModules` to compile files, which basically means you can then only use a subset of the TypeScript language. I couldn't find any open issues against it so I'm not sure if there's any workaround. This likely opens up the debate about what the *right* TypeScript compiler flags..

**Comment by threepointone** [comment]
Hmm it's because we use babel-plugin-typescript, so isolatedModules is enforced. Lemme think about it.

**Comment by LukeSheard** [comment]
> it's because we use babel-plugin-typescript, so isolatedModules is enforced

Yeah - this was one thing that always frustrated me about the babel TypeScript eco-system. 

I think the easiest issue might be to patch the typings themselves - maybe we can add some kind of patch on the node_modules to work around since it's not our code. We can't use the official babel workaround see [1] point 2. 

[1] https://babeljs.io/docs/en/babel-plugin-transform-typescript#caveats

File: package.json [hunk 11]
```diff
@@ -66,6 +66,7 @@
     "*.{js,json,ts,tsx,css,md,mdx}": "prettier --write"
   },
   "resolutions": {
-    "rollup": "^2.33.3"
+    "rollup": "^2.33.3",
+    "**/update-notifier/boxen": "^5.0.0"
```
Comment (threepointone) [inline package.json:70]: why not wait for your update-notifier PR to land?

File: packages/modular-scripts/package.json (hunk 7 above)
Comment (LukeSheard) [reply packages/modular-scripts/package.json:12]: This is just useful if you're hanging out in the `modular-scripts` folder itself rather than needing two prompts for each place in the repo.

File: package.json (hunk 11 above)
Comment (LukeSheard) [reply package.json:70]: I think the real fix we need is in the DefinitelyTyped PR which currently has CI failing - so this is a workaround for now to get them all in line.

**Review by threepointone — APPROVED** [review]
